File: day_6/Lanternfish.cpp
#include <iostream>
#include <fstream>
#include <array>
#include <sstream>
#include <vector>
#include <string>
#include <numeric>

std::vector<int> parseCommaSeparedFileNumbers();
uint64_t solvePartOneTwo(std::vector<int>& initialState, int days);

int main() {
    std::vector<int> initialState = parseCommaSeparedFileNumbers();

    std::cout << "Result <PartOne>: " << solvePartOneTwo(initialState, 80) << std::endl;
    std::cout << "Result <PartTwo>: " << solvePartOneTwo(initialState, 256) << std::endl;
}

uint64_t solvePartOneTwo(std::vector<int>& initialState, int days) {
    std::array<uint64_t, 9> lanternfish { 0 };

    for(auto i: initialState)
        ++lanternfish[i];

    for(size_t day = 0; day < days; day++) 
        lanternfish[(day + 7) % 9] += lanternfish[day % 9];

    return std::accumulate(lanternfish.begin(), lanternfish.end(), 0ll); 
}

std::vector<int> parseCommaSeparedFileNumbers() {
    std::string line;
    std::vector<int> file;
    std::ifstream myfile("input.txt");
    while(getline(myfile, line)) {
        std::stringstream ss(line);
        for (int value; ss >> value; ) {
            file.push_back(value);
            if(ss.peek() == ',') ss.ignore();
        }
    }
    myfile.close();
    return file;
}

File: day_6/Lanternfish.py
from collections import Counter

with open("day_6/input.txt") as file:
    initialState = [int(value) for value in file.readline().split(',')]

state = []
state.append([value -1 for value in initialState])

def solvePartOneTwo(days: int) -> int:
    ''' Works for both parts and it's better '''
    lanternfish =  Counter(initialState) # can use a defaultdict and sum values from initialState
    for day in range(days):
        lanternfish[(day + 7) % 9] += lanternfish[day % 9] # idea from a guy in megathread
        # new_fish = lanternfish[0]
        # lanternfish[0] = lanternfish[1]
        # lanternfish[1] = lanternfish[2]
        # lanternfish[2] = lanternfish[3]
        # lanternfish[3] = lanternfish[4]
        # lanternfish[4] = lanternfish[5]
        # lanternfish[5] = lanternfish[6]
        # lanternfish[6] = lanternfish[7]
        # lanternfish[7] = lanternfish[8]
        # lanternfish[6] += new_fish
        # lanternfish[8] = new_fish
    return sum(lanternfish.values())
    
def main() -> None:
    # print("Result <PartOne>: ", partOneBadSolution(80))
    print("Result <PartOne>: ", solvePartOneTwo(80))
    print("Result <PartTwo>: ", solvePartOneTwo(256))

def partOneBadSolution(days: int) -> int:
    ''' 
        That shit works and it's possible to visualize the generations, but it's still SLOW {AF} 
        DO NOT USE DAYS > 100
    '''
    for _ in range(1, days):
        for line in state:
            states = []
            for value in line:
                if value == 0:
                    states.append(8)
                    states.append(6)
                else:
                    states.append(value - 1)
        state.append(states)
    return len(state[-1])

if __name__ == "__main__":
    main()
